File: companion/bolumler/coklu-regresyon/ornek-01/cozum.R
args <- commandArgs(trailingOnly = TRUE)
if (any(!args %in% c("--check", "--grafik"))) stop("Bilinmeyen secenek")
veri <- read.csv("veri.csv", check.names = FALSE)
yeni <- read.csv("yeni.csv", check.names = FALSE)
kontrol <- function(tablo, sutunlar, en_az) {
  stopifnot(identical(names(tablo), sutunlar), nrow(tablo) >= en_az)
  stopifnot(all(vapply(tablo, is.numeric, logical(1))), all(is.finite(as.matrix(tablo))))
  stopifnot(all(tablo$saat >= 0), all(tablo$devamsizlik >= 0))
  stopifnot(all(tablo$devamsizlik == floor(tablo$devamsizlik)))
}
kontrol(veri, c("saat", "devamsizlik", "puan"), 4)
kontrol(yeni, c("saat", "devamsizlik"), 1)
stopifnot(nrow(yeni) == 1, length(unique(veri$puan)) > 1)
tasarim <- model.matrix(~ saat + devamsizlik, data = veri)
stopifnot(qr(tasarim)$rank == 3, kappa(tasarim, exact = TRUE) <= 1e12)
model <- lm(puan ~ saat + devamsizlik, data = veri, na.action = na.fail)
ozet <- summary(model)
artik <- residuals(model)
sse <- sum(artik^2)
sst <- sum((veri$puan - mean(veri$puan))^2)
stopifnot(sse > .Machine$double.eps * sst)
hacim <- nrow(veri)
serbestlik <- df.residual(model)
mse <- sse / serbestlik
kritik <- qt(.975, serbestlik)
ekle <- function(grup, degerler) {
  data.frame(degisken = grup, olcu = names(degerler), deger = unname(degerler), row.names = NULL)
}
sonuc <- ekle("model", c(n = hacim, serbestlik = serbestlik, sse = sse, sst = sst,
  mse = mse, artik_sd = sqrt(mse), r_kare = ozet$r.squared,
  duzeltilmis_r_kare = ozet$adj.r.squared, f = unname(ozet$fstatistic[1]),
  p_f = pf(unname(ozet$fstatistic[1]), 2, serbestlik, lower.tail = FALSE), alfa = .05, kritik_t = kritik))
rownames(sonuc) <- NULL
araliklar <- confint(model, level = .95)
for (sira in seq_len(3)) {
  grup <- c("sabit", "saat", "devamsizlik")[sira]
  degerler <- unname(coef(ozet)[sira, ])
  sonuc <- rbind(sonuc, ekle(grup, c(katsayi = degerler[1], se = degerler[2],
    t = degerler[3], p = degerler[4], alt = araliklar[sira, 1], ust = araliklar[sira, 2])))
}
yardimci_saat <- lm(saat ~ devamsizlik, data = veri)
yardimci_devam <- lm(devamsizlik ~ saat, data = veri)
sonuc <- rbind(sonuc, ekle("vif", c(saat = 1 / (1 - summary(yardimci_saat)$r.squared),
  devamsizlik = 1 / (1 - summary(yardimci_devam)$r.squared))))
ortalama <- predict(model, newdata = yeni, interval = "confidence", level = .95)
birey <- predict(model, newdata = yeni, interval = "prediction", level = .95)
se_ortalama <- unname(predict(model, newdata = yeni, se.fit = TRUE)$se.fit)
sonuc <- rbind(sonuc, ekle("yeni", c(saat = yeni$saat, devamsizlik = yeni$devamsizlik,
  ongoru = ortalama[1, 1], h_yeni = se_ortalama^2 / mse, se_ortalama = se_ortalama,
  se_birey = sqrt(mse + se_ortalama^2), ortalama_alt = ortalama[1, 2],
  ortalama_ust = ortalama[1, 3], birey_alt = birey[1, 2], birey_ust = birey[1, 3])))
for (sira in seq_len(hacim)) {
  sonuc <- rbind(sonuc, ekle(sprintf("satir_%02d", sira), c(
    uydurulan = unname(fitted(model)[sira]), artik = unname(artik[sira]))))
}
stopifnot(all(is.finite(sonuc$deger)))
print(sonuc, row.names = FALSE, digits = 12)
if ("--check" %in% args) {
  beklenen <- read.csv("beklenen-sonuclar.csv", check.names = FALSE)
  stopifnot(identical(names(beklenen), names(sonuc)))
  stopifnot(identical(sonuc[c("degisken", "olcu")], beklenen[c("degisken", "olcu")]))
  stopifnot(is.numeric(beklenen$deger), all(is.finite(beklenen$deger)))
  p_kontrol <- sonuc$olcu %in% c("p", "p_f")
  tolerans <- ifelse(p_kontrol, 1e-8 * abs(beklenen$deger), 1e-9 + 1e-9 * abs(beklenen$deger))
  stopifnot(all(abs(sonuc$deger - beklenen$deger) <= tolerans))
  cat("DOGRULANDI:", nrow(sonuc), "kontrol degeri eslesiyor.\n")
}
if ("--grafik" %in% args) {
  dir.create("ciktilar/r", recursive = TRUE, showWarnings = FALSE)
  png("ciktilar/r/coklu-regresyon-tani.png", width = 1200, height = 500)
  par(mfrow = c(1, 2))
  plot(fitted(model), artik, xlab = "Uydurulan puan", ylab = "Artik")
  abline(h = 0, lty = 2)
  qqnorm(artik)
  qqline(artik)
  dev.off()
}
for (sutun in c("saat", "devamsizlik")) {
  if (yeni[[sutun]] < min(veri[[sutun]]) || yeni[[sutun]] > max(veri[[sutun]])) {
    warning(paste("Ekstrapolasyon:", sutun))
  }
}
sessionInfo()

File: companion/bolumler/coklu-regresyon/ornek-01/cozum.py
import argparse
import math
from pathlib import Path
import platform

import numpy as np
import pandas as pd
import scipy
from scipy import stats


ALFA = 0.05


def sayisal_kontrol(veri, sutunlar, en_az):
    if list(veri.columns) != sutunlar or len(veri) < en_az:
        raise ValueError("Sutun duzeni veya satir sayisi gecersiz.")
    for sutun in sutunlar:
        if not pd.api.types.is_numeric_dtype(veri[sutun]) or pd.api.types.is_bool_dtype(veri[sutun]):
            raise ValueError("Sayisal sutunlar gerekli.")
    if not np.isfinite(veri.to_numpy(dtype=float)).all():
        raise ValueError("Eksik veya sonsuz deger kabul edilmez; satirlar sessizce silinmez.")
    if (veri.saat < 0).any() or (veri.devamsizlik < 0).any():
        raise ValueError("Saat ve devamsizlik negatif olamaz.")
    if not np.equal(veri.devamsizlik, np.floor(veri.devamsizlik)).all():
        raise ValueError("Devamsizlik tam gun sayisi olmali.")


def hazirla(veri, yeni):
    sayisal_kontrol(veri, ["saat", "devamsizlik", "puan"], 4)
    sayisal_kontrol(yeni, ["saat", "devamsizlik"], 1)
    if len(yeni) != 1:
        raise ValueError("Tek bir yeni tasarim noktasi gerekli.")
    tasarim = np.column_stack([np.ones(len(veri)), veri.saat, veri.devamsizlik]).astype(float)
    if np.linalg.matrix_rank(tasarim) != 3 or np.linalg.cond(tasarim) > 1e12:
        raise ValueError("Tasarim tam rankli ve sayisal olarak iyi kosullu olmali.")
    if veri.puan.nunique() < 2:
        raise ValueError("Sabit puanla bu cikarim paketi kullanilamaz.")
    return veri.reset_index(drop=True).copy(deep=True), yeni.reset_index(drop=True).copy(deep=True), tasarim


def model_hesapla(veri, yeni):
    temiz, nokta, tasarim = hazirla(veri, yeni)
    puan = temiz.puan.to_numpy(dtype=float)
    ortogonal, ust_ucgen = np.linalg.qr(tasarim, mode="reduced")
    katsayi = np.linalg.solve(ust_ucgen, ortogonal.T @ puan)
    ters_ucgen = np.linalg.solve(ust_ucgen, np.eye(3))
    kov_taban = ters_ucgen @ ters_ucgen.T
    uydurulan = tasarim @ katsayi
    artik = puan - uydurulan
    sse = float(artik @ artik)
    merkez = puan - puan.mean()
    sst = float(merkez @ merkez)
    if not math.isfinite(sse + sst) or sse <= np.finfo(float).eps * sst:
        raise ValueError("Tam veya sayisal olarak tama yakin uyumda cikarim hesaplanmaz.")
    hacim = len(temiz)
    serbestlik = hacim - 3
    mse = sse / serbestlik
    se = np.sqrt(mse * np.diag(kov_taban))
    t_degerleri = katsayi / se
    p_degerleri = 2 * stats.t.sf(np.abs(t_degerleri), serbestlik)
    kritik = float(stats.t.ppf(1 - ALFA / 2, serbestlik))
    r_kare = 1 - sse / sst
    f_degeri = max(0.0, (sst - sse) / 2 / mse)
    aciklayicilar = temiz[["saat", "devamsizlik"]].to_numpy(dtype=float)
    rho = float(np.corrcoef(aciklayicilar, rowvar=False)[0, 1])
    vif = 1 / (1 - rho ** 2)
    yeni_tasarim = np.array([1, nokta.saat.iloc[0], nokta.devamsizlik.iloc[0]], dtype=float)
    ongoru = float(yeni_tasarim @ katsayi)
    h_yeni = float(yeni_tasarim @ kov_taban @ yeni_tasarim)
    se_ortalama = math.sqrt(mse * h_yeni)
    se_birey = math.sqrt(mse * (1 + h_yeni))
    satirlar = [("model", isim, deger) for isim, deger in {
        "n": hacim, "serbestlik": serbestlik, "sse": sse, "sst": sst,
        "mse": mse, "artik_sd": math.sqrt(mse), "r_kare": r_kare,
        "duzeltilmis_r_kare": 1 - (1 - r_kare) * (hacim - 1) / serbestlik,
        "f": f_degeri, "p_f": float(stats.f.sf(f_degeri, 2, serbestlik)),
        "alfa": ALFA, "kritik_t": kritik}.items()]
    for indis, ad in enumerate(["sabit", "saat", "devamsizlik"]):
        olculer = {"katsayi": katsayi[indis], "se": se[indis], "t": t_degerleri[indis],
                   "p": p_degerleri[indis], "alt": katsayi[indis] - kritik * se[indis],
                   "ust": katsayi[indis] + kritik * se[indis]}
        satirlar.extend((ad, isim, deger) for isim, deger in olculer.items())
    satirlar.extend([("vif", "saat", vif), ("vif", "devamsizlik", vif)])
    olculer = {"saat": float(nokta.saat.iloc[0]), "devamsizlik": float(nokta.devamsizlik.iloc[0]),
               "ongoru": ongoru, "h_yeni": h_yeni, "se_ortalama": se_ortalama,
               "se_birey": se_birey, "ortalama_alt": ongoru - kritik * se_ortalama,
               "ortalama_ust": ongoru + kritik * se_ortalama,
               "birey_alt": ongoru - kritik * se_birey, "birey_ust": ongoru + kritik * se_birey}
    satirlar.extend(("yeni", isim, deger) for isim, deger in olculer.items())
    for indis, (uyum, hata) in enumerate(zip(uydurulan, artik), start=1):
        satirlar.extend([(f"satir_{indis:02}", "uydurulan", uyum), (f"satir_{indis:02}", "artik", hata)])
    sonuc = pd.DataFrame(satirlar, columns=["degisken", "olcu", "deger"])
    if not np.isfinite(sonuc.deger).all():
        raise ValueError("Sonlu model sonuclari hesaplanamadi.")
    return sonuc, uydurulan, artik


def hesapla(veri, yeni):
    return model_hesapla(veri, yeni)[0]


def kontrol_et(sonuc, beklenen):
    if list(beklenen.columns) != ["degisken", "olcu", "deger"]:
        raise ValueError("Kontrol sutunlari uyusmuyor.")
    if not sonuc[["degisken", "olcu"]].equals(beklenen[["degisken", "olcu"]]):
        raise ValueError("Kontrol etiketleri veya sirasi uyusmuyor.")
    hedefler = pd.to_numeric(beklenen.deger, errors="raise")
    if not np.isfinite(hedefler).all():
        raise ValueError("Sonlu referanslar gerekli.")
    for olcu, gercek, hedef in zip(sonuc.olcu, sonuc.deger, hedefler):
        p_kontrol = olcu in {"p", "p_f"}
        if not math.isclose(gercek, hedef, rel_tol=1e-8 if p_kontrol else 1e-9,
                            abs_tol=0 if p_kontrol else 1e-9):
            raise ValueError(f"Kontrol uyusmazligi: {olcu}: {gercek} != {hedef}")
    print(f"DOGRULANDI: {len(sonuc)} kontrol degeri eslesiyor.")


def grafik_kaydet(veri, yeni):
    import matplotlib
    matplotlib.use("Agg")
    import matplotlib.pyplot as plt

    sonuc, uydurulan, artik = model_hesapla(veri, yeni)
    fig, eksenler = plt.subplots(1, 2, figsize=(10, 4))
    eksenler[0].scatter(uydurulan, artik, color="#1F4E79")
    eksenler[0].axhline(0, color="#555555", linestyle="--")
    eksenler[0].set(xlabel="Uydurulan puan", ylabel="Artik", title="Artik - uydurulan")
    stats.probplot(artik, dist="norm", plot=eksenler[1])
    eksenler[1].set(title="Normal Q-Q", xlabel="Kuramsal normal kantil", ylabel="Sirali artik")
    fig.tight_layout()
    hedef = Path("ciktilar/python")
    hedef.mkdir(parents=True, exist_ok=True)
    fig.savefig(hedef / "coklu-regresyon-tani.png", dpi=150)
    plt.close(fig)


def main():
    parser = argparse.ArgumentParser(description="Kapsamli bolum 15: coklu dogrusal regresyon")
    parser.add_argument("--check", action="store_true")
    parser.add_argument("--grafik", action="store_true")
    secenek = parser.parse_args()
    veri, yeni = pd.read_csv("veri.csv"), pd.read_csv("yeni.csv")
    sonuc = hesapla(veri, yeni)
    print(sonuc.to_string(index=False))
    if secenek.check:
        kontrol_et(sonuc, pd.read_csv("beklenen-sonuclar.csv"))
    if secenek.grafik:
        grafik_kaydet(veri, yeni)
    for sutun in ["saat", "devamsizlik"]:
        if not veri[sutun].min() <= yeni[sutun].iloc[0] <= veri[sutun].max():
            print(f"UYARI: yeni {sutun} gozlenen araligin disinda; ekstrapolasyon.")
    print(f"Python {platform.python_version()}, NumPy {np.__version__}, pandas {pd.__version__}, SciPy {scipy.__version__}")


if __name__ == "__main__":
    main()